internal/queue/ctqueue.go

```diff
@@ -0,0 +1,120 @@
+package queue
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// CTQueue is a thread safe queue implementation which allows safe usage among different threads/go-routines and can
+// be closed at any given time. The queue can be closed by either allowing the remaining elements to remain within or
+// to be extracted. Once the queue is closed, no new elements may be added, but existing elements can still be popped of
+// the queue.
+type CTQueue[T any] struct {
```

Comment:
PChan had some unit tests which caught a few issues in the past. I'd feel more comfortable if this type had some unit tests as well.

Reply:
Done